Add plugin: FLARE.ai

AlexXYK opened the pull request:
# I am submitting a new Community Plugin

## Repo URL

<!--- Paste a link to your repo here for easy access -->
Link to my plugin: https://github.com/AlexXYK/flare-ai/

## Release Checklist
- [x] I have tested the plugin on
  - [x]  Windows
  - [ ]  macOS
  - [x]  Linux
  - [x]  Android _(if applicable)_
  - [x]  iOS _(if applicable)_
- [x] My GitHub release contains all required files (as individual files, not just in the source.zip / source.tar.gz)
  - [x] `main.js`
  - [x] `manifest.json`
  - [x] `styles.css` _(optional)_
- [x] GitHub release name matches the exact version number specified in my manifest.json (_**Note:** Use the exact version number, don't include a prefix `v`_)
- [x] The `id` in my `manifest.json` matches the `id` in the `community-plugins.json` file.
- [x] My README.md describes the plugin's purpose and provides clear usage instructions.
- [x] I have read the developer policies at https://docs.obsidian.md/Developer+policies, and have assessed my plugins's adherence to these policies.
- [x] I have read the tips in https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines and have self-reviewed my plugin to avoid these common pitfalls.
- [x] I have added a license in the LICENSE file.
- [x] My project respects and is compatible with the original license of any code from other plugins that I'm using.
      I have given proper attribution to these other projects in my `README.md`.

AlexXYK commented:
Changes pushed, including optionals - v1.01 release.

joethei commented:
["icon": "flame",](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/manifest.json#L10)
This is not a thing in the manifest, remove this.

[async activateView(location: 'tab' | 'right' | 'left' = 'tab') {](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/main.ts#L228-L286)
Just offer one command that opens the view in the sidebar, then let the user move that manually.
No need to overly convolute this code here.

[private parseFrontmatter(content: string): any {](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/main.ts#L1024-L1042). [const frontmatter: FrontMatter = {};](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/history/ChatHistoryManager.ts#L283-L295), [// Extract frontmatter and system prompt](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/flares/FlareManager.ts#L136-L161), [private parseFrontmatter(content: string): any {](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/flares/FlareManager.ts#L211-L224), [private async processFrontmatter(content: string): Promise<any> {](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/flares/FlareManager.ts#L239-L261), [const lines = content.split('\n');](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/flares/FlareManager.ts#L1631-L1678)
Use the MetadataCache API to access frontmatter, instead of rolling your own YAML parsing code.

[.slice(0, 10); // Limit to 10 results for performance](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L171), [.slice(0, 10);](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L150)
Instead of slicing use the `limit` property that is available on the SuggestModal class.

[container.setAttribute('aria-label', 'FLARE.ai Chat Interface');](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L447), [chatContainer.setAttribute('aria-label', 'Chat Messages');](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L452), [mainContent.setAttribute('aria-label', 'Main Chat Area');](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L457), [toolbar.setAttribute('aria-label', 'Chat Controls');](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L462), [toolbarCenter.createEl('h2', { text: 'New Chat' });](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L482), [titleEl.textContent = 'New Chat';](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L1063), [.setName('API Key')](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/components/ProviderSettingsView.ts#L99), [.setName('API Key')](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/components/ProviderSettingsView.ts#L117), [.setName('API Key')](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/components/ProviderSettingsView.ts#L147), [.setName('API Key')](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/components/ProviderSettingsView.ts#L190), ['New Chat',](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/components/HistorySidebar.ts#L518), ['Chat History'](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/components/HistoryModal.ts#L32), [text: 'New Chat'](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/components/HistoryModal.ts#L45), [.setName('Provider Name')](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/providers/ProviderManager.ts#L125), [.setName('Enable Provider')](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/providers/ProviderManager.ts#L137), [.setName('API Key')](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/providers/ProviderManager.ts#L237), [.setName('API Key')](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/providers/ProviderManager.ts#L271), [.setName('Available Models')](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/providers/ProviderManager.ts#L310), [this.titleEl.setText('Confirm Resend');](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/modals/ConfirmModal.ts#L14), [title: title || 'New Chat',](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/history/ChatHistoryManager.ts#L51), [.setName('Flare Name')](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/flares/FlareManager.ts#L551), [.setName('Reasoning Model')](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/flares/FlareManager.ts#L611), [.setName('Reasoning Header')](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/flares/FlareManager.ts#L644), [.setName('System Prompt')](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/flares/FlareManager.ts#L716), [.setName('Context Window')](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/flares/FlareManager.ts#L727), [.setName('Handoff Context')](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/flares/FlareManager.ts#L741), [.setName('Default Temperature')](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/flares/FlareManager.ts#L755), [.setButtonText('Delete Flare')](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/flares/FlareManager.ts#L770), ['Delete Flare',](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/flares/FlareManager.ts#L882), [.setName('Reasoning Model')](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/flares/FlareManager.ts#L1105), [.setName('Reasoning Header')](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/flares/FlareManager.ts#L1122), [.setName('Reasoning Header')](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/flares/FlareManager.ts#L1147), [.setName('Enable Streaming')](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/flares/FlareManager.ts#L1253), [.setName('Max Tokens')](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/flares/FlareManager.ts#L1292), [.setName('Context Window')](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/flares/FlareManager.ts#L1313), [.setName('Handoff Context')](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/flares/FlareManager.ts#L1334), ['System Prompt'](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/flares/FlareManager.ts#L1359), ['Unsaved Changes',](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/flares/FlareManager.ts#L1408)
[Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)

[container.setAttribute('aria-label', 'FLARE.ai Chat Interface');](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L447), [chatContainer.setAttribute('aria-label', 'Chat Messages');](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L452), [mainContent.setAttribute('aria-label', 'Main Chat Area');](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L457), [toolbar.setAttribute('aria-label', 'Chat Controls');](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L462), [attr: { 'aria-label': 'New chat' }](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L477), [attr: { 'aria-label': 'Toggle chat history' }](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L470), [attr: { 'aria-label': 'Save chat' }](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L490), [attr: { 'aria-label': 'Clear chat' }](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L497), [attr: { 'aria-label': 'Choose flare' }](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L513), [attr: { 'aria-label': 'Send message', type: 'button' }](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L543) and more
Use the `setTooltip` function from the API instead.
(I don't want to see any more uses of `aria-label` when I rereview this plugin

[await MarkdownRenderer.renderMarkdown(responsePart.trim(), rendered, '', this.plugin);](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L1421), [await MarkdownRenderer.renderMarkdown(message.content, rendered, '', this.plugin);](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L1427), [await MarkdownRenderer.renderMarkdown(message.content, rendered, '', this.plugin);](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L1433), [await MarkdownRenderer.renderMarkdown(joinedReasoning, reasoningContent, '', this.plugin);](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L1862), [await MarkdownRenderer.renderMarkdown(content, rendered, '', this.plugin);](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L2346), [await MarkdownRenderer.renderMarkdown(joinedReasoning, reasoningContainer, '', this.plugin);](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L2371), [await MarkdownRenderer.renderMarkdown(responsePart.trim(), responseContainer, '', this.plugin);](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L2378)
This function is deprecated, use the newer `MarkdownRenderer.render` function.
This will also fix the console errors that are currently generated by your plugin.

[private createDebouncedHandler(handler: () => void, delay: number, immediate: boolean) {](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/aiChatView.ts#L1470-L1477), [// Add debouncing for flare loading](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/flares/FlareManager.ts#L318-L345)
Use the `debounce` function from the Obsidian API instead of rolling your own implementation.

[const generalSection = this.createSection(settingsContainer, 'General');](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/components/ProviderSettingsView.ts#L24), [containerEl.createEl('h1', { text: 'FLARE.ai' });](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/settings/GeneralSettingTab.ts#L27), [const generalSection = this.createSection(wrapper, 'General', true);](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/settings/GeneralSettingTab.ts#L40)
[Don't add a top-level heading in the settings tab, such as "General", "Settings", or the name of your plugin.](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Only+use+headings+under+settings+if+you+have+more+than+one+section.)

[if (Platform.isMobile) {](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/components/ProviderSettingsView.ts#L210-L212)
This is not necessary, the `is-mobile` class is already in the dom, in the `body` element.

[private showContextMenu(item: HTMLElement, x: number, y: number) {](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/components/HistorySidebar.ts#L490-L528), [private addContextMenuItem(text: string, icon: string, onClick: () => void, isDanger = false) {](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/views/components/HistorySidebar.ts#L530-L547)
Use the `Menu` class from Obsidian instead of building your own

For section headings in settings use:
```ts
new Setting(containerEl).setName('name here').setHeading();
```
[header.createSpan({ cls: 'flare-section-chevron' });](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/settings/GeneralSettingTab.ts#L56-L57)

[if (collapsible) {](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/settings/GeneralSettingTab.ts#L59-L64)
No settings hidden behind a collapsed section please.

[const response = await fetch(`${this.url}/models`, {](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/providers/implementations/OpenRouter/OpenRouterProvider.ts#L57), [const response = await fetch(`${this.url}/chat/completions`, {](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/providers/implementations/OpenRouter/OpenRouterProvider.ts#L131), [const response = await fetch('https://api.openai.com/v1/chat/completions', {](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/providers/implementations/OpenAI/OpenAIProvider.ts#L105), [const response = await fetch(`${this.config.baseUrl}/api/chat`, {](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/providers/implementations/Ollama/OllamaProvider.ts#L99)
Use the `requestUrl` function from the Obsidian API instead, it will handle some things like CORS automatically.


[await this.plugin.app.vault.modify(file, content);](https://github.com/AlexXYK/flare-ai/blob/bc13c7beaf0839c0e8b20487d699bca27df134d6/src/flares/FlareManager.ts#L1457)
To edit frontmatter in an already existing node use the `processFrontMatter` API, no

AlexXYK commented:
@joethei I have updated this plugin with some new features, and also addressed the issues you mentioned above; used as many Obsidian-native APIs as possible. Note, I still use fetch for the streaming features with OpenAI and OpenRouter providers. Please let me know if there is anything else that can be fixed.

AlexXYK commented:
tag corrected.

AlexXYK commented:
PR conflict fixed.

joethei commented:
[  new Setting(generalSection).setName('General').setHeading();](https://github.com/AlexXYK/flare-ai/blob/899b4935f02bf940c721c2f2490360753139653a/src/settings/GeneralSettingTab.ts#L43)
General settings should be at the top and should not have a header.

AlexXYK commented:
General settings moved.
